File: AlgoExpert/Graphs/Medium/Minimum Passes Of Matrix.java
import java.util.*;

class Program {

  public int minimumPassesOfMatrix(int[][] matrix) {
    // Write your code here.
    int current = 0;
    int next = 1;
    int numOfPasses = 0;
    Queue<Integer[]> queue = new LinkedList<>();

    for (int row = 0; row < matrix.length; row++) {
      for (int column = 0; column < matrix[row].length; column++) {
        if (matrix[row][column] > 0)
          queue.add(new Integer[] {row, column, current});
      }
    }

    while (!queue.isEmpty()) {
      Integer[] currentPos = queue.poll();
      int row = currentPos[0];
      int column = currentPos[1];
      if (currentPos[2] == next) {
        numOfPasses++;
        next = current;
        current = currentPos[2];
      }
      if (row > 0 && matrix[row - 1][column] < 0) {
        matrix[row - 1][column] *= -1;
        queue.add(new Integer[] {row - 1, column, next});
      }
      if (row < matrix.length - 1 && matrix[row + 1][column] < 0) {
        matrix[row + 1][column] *= -1;
        queue.add(new Integer[] {row + 1, column, next});
      }
      if (column > 0 && matrix[row][column - 1] < 0) {
        matrix[row][column - 1] *= -1;
        queue.add(new Integer[] {row, column - 1, next});
      }
      if (column < matrix[row].length - 1 && matrix[row][column + 1] < 0) {
        matrix[row][column + 1] *= -1;
        queue.add(new Integer[] {row, column + 1, next});
      }
    }

    for (int row = 0; row < matrix.length; row++) {
      for (int column = 0; column < matrix[row].length; column++) {
        if (matrix[row][column] < 0)
          return -1;
      }
    }
    
    return numOfPasses;
  }
}

File: Python/AlgoExpert/Graphs/Medium/minimumPassesOfMatrix.py
def minimumPassesOfMatrix(matrix):
    # Write your code here.
    passes = 0
    valid_negatives, has_negatives = find_valid_negatives(matrix)
    while len(valid_negatives) > 0:
        for negative in valid_negatives:
            row = negative[0]
            col = negative[1]
            matrix[row][col] *= -1
        passes += 1
        valid_negatives, has_negatives = find_valid_negatives(matrix)
    
    if not has_negatives:
        return passes
    else:
        return -1


def find_valid_negatives(matrix):
    negatives = []
    has_negatives = False
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if matrix[i][j] < 0:
                if has_adjacent_positives(matrix, i, j):
                    negatives.append((i, j))
                has_negatives = True
    return negatives, has_negatives


def has_adjacent_positives(matrix, row, col):
    if row - 1 >= 0 and matrix[row - 1][col] > 0:
        return True
    elif row + 1 < len(matrix) and matrix[row + 1][col] > 0:
        return True
    elif col - 1 >= 0 and matrix[row][col - 1] > 0:
        return True
    elif col + 1 < len(matrix[row]) and matrix[row][col + 1] > 0:
        return True
    
    return False


    